solutions/noisg-19-feast.mdx

```diff
@@ -0,0 +1,81 @@
+---
+id: noisg-19-feast
+source: NOI
+title: 2019 - Feast
+author: Dong Liu
+---
+
+**Time Complexity:** $\mathcal{O}(N\log{\sum a[i]})$.
+
+<LanguageSection>
+
+<CPPSection>
+
+Like a normal Alien Trick problem, we're going to binary search on the maximum $\texttt{penalty}$.
+
+Now, for a given penalty, we're going to calculate the number of people in an optimal food arrangement such that we subtract $\texttt{penalty}$ for every person.
+
+Lets have $\texttt{dp}[i][j:\{0,1\}]$ represent the maximum sum of satisfaction of an arrangement of the first $i$ people, given that $j=0/1$ implies whether person $i$ is being used. Let $\texttt{cnt}[i][j]$ represent the number of people used in an optimal arrangement of $\texttt{dp}[i][j]$.  
+
+For our $\texttt{dp}$ transitions, we have
+
+$$
+\{\texttt{dp}[i][0], \texttt{cnt}[i][0]\} = 
+\max\begin{cases} \{\texttt{dp}[i - 1][0], \texttt{cnt}[i - 1][0]\}\\
+\{\texttt{dp}[i - 1][1], \texttt{cnt}[i - 1][1]\}
+\end{cases}
+
+$$
+and 
+$$
+\{\texttt{dp}[i][1], \texttt{cnt}[i][1]\} = 
+\max\begin{cases}\{\texttt{dp}[i - 1][0] + a[i] - \texttt{penalty}, \texttt{cnt}[i - 1][0] + 1\}\\
+\{\texttt{dp}[i - 1][1] + a[i], \texttt{cnt}[i - 1][1]\}\end{cases}
+$$
+
+because we either begin a new subarray or we continue an existing subarray.
+
+Since calculating $\texttt{dp}$ takes $\mathcal O(N)$, and we're binary searching on the range $[0, \sum a[i]]$, the time complexity is $\mathcal{O}(N\log{\sum a[i]})$.
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+const int N = 300000;
+const long long INF = (long long) 300000 * 1000000000;
+
+int n, k, a[N + 1];
+pair<long double, int> dp[N + 1][2];
+	
+bool check(long double penalty) {
+	dp[0][0] = {0, 0}, dp[0][1] = {-INF, 0};
+	for (int i = 1; i <= n; ++i) {
+		dp[i][0] = max(dp[i - 1][0], dp[i - 1][1]);
+		dp[i][1] = max(make_pair(dp[i - 1][1].first + a[i], dp[i - 1][1].second),
+			make_pair(dp[i - 1][0].first + a[i] - penalty, dp[i - 1][0].second + 1));
+	}
+	return max(dp[n][0], dp[n][1]).second >= k;
+};
+
+int main() {
+	scanf("%d%d", &n, &k);
+	for (int i = 1; i <= n; ++i)
+		scanf("%d", a + i);
+	long double lower = 0, upper = INF;
+	for (int it = 0; it < 100; ++it) {
```

Comment:
why only binsearch 100 times

Reply:
binary searching 100 times is precise enough

Reply:
but like just binary search until lower and upper meet up? why not do that?

Reply:
sure

Reply:
https://oj.uz/submission/440784 rip oj.uz queue

Reply:
i don't get it

Reply:
it was "in queue" for half a day